docker.py

```diff
@@ -193,22 +194,31 @@ def load(readable):
     return loaded_image
 
 
-def build_and_tag(tag_name, dockerfile_path, cwd=None, log_path=None):
-    print("Building runner Docker image %s from %s" % (tag_name,
-                                                       dockerfile_path))
+def build_and_tag(tag_name, dockerfile_path, cwd=None,
+                  log_path=None, args=None):
+    if args is None:
+        args = {}
+    build_args = list(reduce(lambda a, b: a + b,
+                             map(lambda k: [
+                                 "--build-arg",
+                                 "%s=%s" % (k, args[k])
+                             ], args), []))
```

Comment:
While I admire the functional approach quite a lot, I think one of the following two is easier to read.
```suggestion
    build_args = []
    for k, v in args.items():
        build_args += ["--build-arg", f"{k}={v}"]
```

```suggestion
    build_args = [e for k, v in args.items()
                  for e in ("--build-args", f"{k}={v}")]
```

(Don't forget to remove the import)